src/circuit/gadget/sinsemilla/merkle.rs

```diff
@@ -175,6 +176,289 @@ impl MerkleChip {
+                        || "copy b",
+                        config.b_a1,
+                        0,
+                        &b.cell_value(),
+                        &config.perm,
+                    )?;
+                    // Copy and assign `c` at the correct position.
+                    copy(
+                        &mut region,
+                        || "copy c",
+                        config.c_b0,
+                        0,
+                        &c.cell_value(),
+                        &config.perm,
+                    )?;
+                    // Copy and assign the left node at the correct position.
+                    copy(
+                        &mut region,
+                        || "left",
+                        config.left_b1,
+                        0,
+                        &left,
+                        &config.perm,
+                    )?;
+                    // Copy and assign the right node at the correct position.
+                    copy(
+                        &mut region,
+                        || "right",
+                        config.right,
+                        0,
+                        &right,
+                        &config.perm,
+                    )?;
+
+                    // Copy and assign the subpiece `a_0` at the correct position.
+                    copy(
+                        &mut region,
+                        || "a_0",
+                        config.a_a0,
+                        1,
+                        &a.subpieces()[0].cell_value(),
+                        &config.perm,
+                    )?;
+                    // Copy and assign the subpiece `a_1` at the correct position.
+                    copy(
+                        &mut region,
+                        || "a_1",
+                        config.b_a1,
+                        1,
+                        &a.subpieces()[1].cell_value(),
+                        &config.perm,
+                    )?;
+                    // Copy and assign the subpiece `b_0` at the correct position.
+                    copy(
+                        &mut region,
+                        || "b_0",
+                        config.c_b0,
+                        1,
+                        &b.subpieces()[0].cell_value(),
+                        &config.perm,
+                    )?;
+                    // Copy and assign the subpiece `b_1` at the correct position.
+                    copy(
+                        &mut region,
+                        || "b_1",
+                        config.left_b1,
+                        1,
+                        &b.subpieces()[1].cell_value(),
+                        &config.perm,
+                    )?;
+
+                    Ok(())
+                },
+            )?;
+        }
+
+        let (point, _) = self.hash_to_point(
+            layouter.namespace(|| format!("l_star {}", l_star)),
+            Q,
+            vec![a, b, c].into(),
```

Comment:
I believe this is where `a`, `b`, and `c` are constrained to be their correct bit ranges. However, their sub-pieces have not been range-constrained anywhere, which means they could actually be witnessed by the prover such that they overflow the additions in the "Check piece decomposition" gate. This would enable the prover to witness different `left` and `right` values than are actually input to Sinsemilla.

In addition to returning the running sums from `SinsemillaInstructions::hash_to_point`, we should also return the assigned `bits` cells for each piece, which are guaranteed (by the lookups) to be 10-bit values. Then instead of allowing the prover to witness `a0` etc, we directly reconstruct `left` and `right` from these pieces. This requires 51 cells in the `Check piece decomposition" region for input pieces instead of 5 (which I'm almost certain is unavoidable - @daira?) plus a 5-bit decomposition (to split one of the 10-bit limbs between `left` and `right`).

Reply:
// `a = a_0||a_1` = `l_star` (10 bits) || (bits 0..=239 of `left`)
// `b = b_0||b_1` = (bits 240..=254 of `left`) || (bits 0..=234 of `right`)
// `c = bits 235..=254 of `right`

i think we can recover the subpieces using fewer cells:
- `a` is witnessed and constrained by `SinsemillaHash(a)` to be 250 bits. `SinsemillaHash(a)` outputs `z_a0`, `z_a1`, ..., `z_a24`, where `z_a0` = `a`
    - `a_1` = `z_a1`
    - `a_0` = `a` - (2^10)`a_1`
- `b` is witnessed and constrained by `SinsemillaHash(b)` to be 250 bits
    - witness `b_0` and manually check that it is a 5-bit value
    - `b_1` = (`b` - `b_0`) / 2^5
- `c` is witnessed and constrained by `SinsemillaHash(c)` to be 20 bits